cli/command/service/progress/progress.go

```diff
@@ -0,0 +1,378 @@
+				})
+			}
+		} else {
+			if !convergedAt.IsZero() {
+				progressOut.WriteProgress(progress.Progress{
+					ID:     "verify",
+					Action: "Detected task failure",
+				})
+			}
+			convergedAt = time.Time{}
+		}
+
+		select {
+		case <-time.After(200 * time.Millisecond):
+		case <-sigint:
+			if !converged {
+				progress.Message(progressOut, "", "Operation continuing in background.")
+				progress.Messagef(progressOut, "", "Use `docker service ps %s` to check progress.", serviceID)
+			}
+			return nil
+		}
+	}
+}
+
+func writeOverallProgress(progressOut progress.Output, numerator, denominator int, rollback bool) {
+	if rollback {
+		progressOut.WriteProgress(progress.Progress{
+			ID:     "overall progress",
+			Action: fmt.Sprintf("rolling back update: %d out of %d tasks", numerator, denominator),
+		})
+		return
+	}
+	progressOut.WriteProgress(progress.Progress{
+		ID:     "overall progress",
+		Action: fmt.Sprintf("%d out of %d tasks", numerator, denominator),
+	})
+}
+
+type replicatedProgressUpdater struct {
+	progressOut progress.Output
+
+	// used for maping slots to a contiguous space
+	// this also causes progress bars to appear in order
+	slotMap map[int]int
+
+	initialized bool
+	done        bool
+}
+
+func (u *replicatedProgressUpdater) update(ctx context.Context, service swarm.Service, tasks []swarm.Task, activeNodes map[string]swarm.Node, rollback bool) (bool, error) {
+	if service.Spec.Mode.Replicated == nil || service.Spec.Mode.Replicated.Replicas == nil {
+		return false, errors.New("no replica count")
+	}
+	replicas := *service.Spec.Mode.Replicated.Replicas
+
+	if !u.initialized {
+		u.slotMap = make(map[int]int)
+
+		// Draw progress bars in order
+		writeOverallProgress(u.progressOut, 0, int(replicas), rollback)
+
+		if replicas <= maxProgressBars {
+			for i := uint64(1); i <= replicas; i++ {
+				progress.Update(u.progressOut, fmt.Sprintf("%d/%d", i, replicas), " ")
+			}
+		}
+		u.initialized = true
+	}
+
+	// If there are multiple tasks with the same slot number, favor the one
+	// with the *lowest* desired state. This can happen in restart
+	// scenarios.
+	tasksBySlot := make(map[int]swarm.Task)
+	for _, task := range tasks {
+		if stateToProgress[task.DesiredState] == 0 {
+			continue
+		}
+		if existingTask, ok := tasksBySlot[task.Slot]; ok {
+			if stateToProgress[existingTask.DesiredState] <= stateToProgress[task.DesiredState] {
+				continue
+			}
+		}
+		tasksBySlot[task.Slot] = task
+	}
+
+progressBars:
+	running := uint64(0)
+
+	for _, task := range tasksBySlot {
+		if _, nodeActive := activeNodes[task.NodeID]; nodeActive {
+			mappedSlot := u.slotMap[task.Slot]
+			if mappedSlot == 0 {
+				mappedSlot = len(u.slotMap) + 1
+				u.slotMap[task.Slot] = mappedSlot
+			}
+			if replicas <= maxProgressBars && uint64(mappedSlot) <= replicas {
+				if u.done {
+					if task.Status.State != swarm.TaskStateRunning {
+						u.done = false
+						goto progressBars
```

Comment:
It does unfortunately require some unusual syntax in go, but otherwise is a pretty common way of doing this.

I think the behaviour of a closure is much better understood.